enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes default `/32` subnet.
+
+##### CRD vs. a labelled / annotated approach
+
+The cloud's node subnet information needs to be communicated to the network
+plugins. Two models have been imagined for this: a CRD based approach, or a node
+labelled / annotated approach.
+
+**CRD**
+
+The cloud component could use a dedicated CRD that the network plugin watches.
+The CRD would store the nodes subnet, as known by the cloud, and perhaps contain
+a status field indicating a successful cloud assignment of the egress IP. Doing
+this would however introduce explicit dependencies between the network plugin
+and this cloud component. If the component ends up being a dedicated OpenShift
+one: the CRD would end up creating a dependency tree between upstream
+OVN-Kubernetes and OpenShift specific components, which is anything but
+desirable.
+
+**labelled / annotated approach**
+
+The second approach imagined is labelling / annotating the Kubernetes node
+object. The cloud component would query the cloud API for the nodes subnet range
+and label / annotate the node objects according to a defined convention which
+the network plugins can read. That information would then be taken into account
+during the plugin's assignment procedure.
+
+Given the minimal information needed to be conveyed to the network plugin from
+the cloud component, and the dependencies that a CRD would introduce: it is
+estimated that the labelled / annotated model might be preferred. The only thing
+not covered by that approach is indicating to the user if the cloud egress IP
+assignment was successful or not. There are other Kubernetes objects however
+that could cover that: events, for example. This model also does not provide a
+certified Kubernetes reconciliation loop, as the CRD model would.
+
+##### Cloud credentials operator - CCO
+
+The cloud credentials operator (CCO) [10] is an OpenShift component that manages
+the `CredentialRequest` CR. This resource allows other components access to the
+cloud API by creating a secret in the `kube-system` namespace which the resource
+issuing the `CredentialRequest` can fetch, containing the permissions required
+to communicate with the cloud API by the requesting component. The utilization
+of the CCO is dependent upon the questions in the [Summary](#summary), as it
+might be awkward deploying the CCO on a non-OpenShift cluster. This would be the
+case if an upstream development model is chosen.
+
+[10]: https://github.com/openshift/cloud-credential-operator
+
+##### RBAC
+
+Given the approach chosen above a separate set of RBAC permissions will be
+required for the cloud component / network plugins. These are listed below.
+
+Independently of the model chosen, the cloud component will require:
+
+WATCH/LIST/GET:
+
+- `egressips.k8s.ovn.org` - to watch for status assignments done by
```

Comment:
Not necessarily; the controller would need the ability to read IP requests. These could just as easily take place entirely via annotations on the Node.

Reply:
Indeed, and that would also solve the problem of issuing "cloud egress IP requests" for openshift-sdn. That is a problem for openshift-sdn since does not indicate a egress IP status, as mentioned in the doc.